CrossSectionAnalysis/cpp/CrossSectionCompute.cxx

```diff
@@ -0,0 +1,340 @@
+        vtkPoints * modelPoints = inputModel->GetMesh()->GetPoints();
+        const unsigned int numberOfPoints = modelPoints->GetNumberOfPoints();
+        if (endPointIndex == (numberOfPoints - 1))
+        {
+            const unsigned int blockLastPointIndex = bufferArray->GetNumberOfTuples() - 1;
+            const unsigned int blockBeforeLastPointIndex = bufferArray->GetNumberOfTuples() - 1 - 1;
+            bufferArray->SetTuple3(blockLastPointIndex,
+                                   bufferArray->GetTuple3(blockLastPointIndex)[0],
+                                   bufferArray->GetTuple3(blockBeforeLastPointIndex)[1],
+                                   bufferArray->GetTuple3(blockBeforeLastPointIndex)[2]);
+        }
+    }
+    
+    // The centerline copy is not needed anymore.
+    inputCenterlineNode->Delete();
+    
+    #if DEVTIME != 0
+    mtx.lock();
+    double endTime = GetTimeOfDay();;
+    cout << "This thread : " << std::this_thread::get_id() << " "
+        << (endTime - startTime) / 1000000 << " seconds" << endl;
+    mtx.unlock();
+    #endif
+}
+
+// Translated and adapted from the Python implementation.
+vtkPolyData * vtkCrossSectionComputeWorker::ComputeCrossSectionPolydata(
+                                vtkMRMLNode * inputCenterlineNode,
+                                vtkPolyData * closedSurfacePolyData,
+                                unsigned int pointIndex)
+{
+    if (inputCenterlineNode == NULL)
+    {
+        WORKER_MESSAGE("Input centerline is NULL.");
+        return NULL;
+    }
+    if (closedSurfacePolyData == NULL)
+    {
+        WORKER_MESSAGE("Closed  surface polydata is NULL.");
+        return NULL;
+    }
+    
+    double center[3] = {0.0, 0.0, 0.0};
+    double normal[3] = {0.0, 0.0, 0.0};
+    if (inputCenterlineNode->IsA("vtkMRMLModelNode"))
+    {
+        vtkMRMLModelNode * inputModel = vtkMRMLModelNode::SafeDownCast(inputCenterlineNode);
+        vtkPoints * modelPoints = inputModel->GetMesh()->GetPoints();
+        const unsigned int numberOfPoints = modelPoints->GetNumberOfPoints();
+        double centerLocal[3] = {0.0, 0.0, 0.0};
+        modelPoints->GetPoint(pointIndex, centerLocal);
+        inputModel->TransformPointToWorld(centerLocal, center);
+        // Exclude the last centerline point
+        if (pointIndex < (numberOfPoints - 1))
+        {
+            double centerInc[3] = {0.0, 0.0, 0.0};
+            // note that this +1 does not work for the last point
+            double centerLocalInc[3] = {0.0, 0.0, 0.0};
+            modelPoints->GetPoint(pointIndex + 1, centerLocalInc);
+            inputModel->TransformPointToWorld(centerLocalInc, centerInc);
+            for (unsigned int i = 0; i < 3; i++)
+                normal[i] = centerInc[i] - center[i];
+        }
+        else
+        // Return an empty polydata at the last centerline point
+        {
+            return vtkPolyData::New();
```

Comment:
Avoid transferring ownership between functions. It is generally much safer and cleaner if you create an object in the same method where you delete that object. In this case, you can create the vtkPolyData instance outside and just pass the pointer as an input argument to the function. The function will update the vtkPolyData content. You can use ShallowCopy (or maybe DeepCopy, but probably that's not needed) to modify an existing polydata with a filter output.